### DISCUSSION TITLE
fork and zip

### DISCUSSION THREAD

**Discussion by hillin** [iter-tools/iter-tools#481] [Q&A] [answered]
I have a pipeline, at a certain point I want to fork it into two copies: one will be fed into a function which maps the iterable into another one; the other will be kept unmodified; after that I'd like to zip the mapped iterator and the unmodified copy. Is there a way to do this in-pipeline?

I have a vague idea that the `fork` operator may be involved, but no idea how to use it in a pipeline.

It should be equivalent to:

```typescript
const source = generateMyIterator();
const converted = someTransformFunction(source);
return map(zip(converted, source), zipThem);
```

I can do:
```typescript
execPipe(
  generateMyIterator(),
  source => [someTransformFunction(source), source],
  // but now how do I zip them?
)
```

**Comment by hillin** [comment]
Well I figured it out:

```typescript
execPipe(
  generateMyIterator(),
  source => zip(someTransformFunction(source), source),
  zipThem
)
```

**Comment by conartist6** [comment, marked as answer]
My best guess is that the correct code is:
```js
map(value => [value, transformValue(value)], generate()); 
```
Alternatively if you really have to transform the whole input at once before you zip, then the correct code would be:
```js
execPipe(
  generate(),
  toArray,
  source => [source, map(transformValue, source)],
  apply(zip),
  toArray,
);
```
[Here is a working example.](https://runkit.com/conartist6/62a36370a2c504000896751e)

**Reply by conartist6** [reply]
In this case `apply(zip)` is more or less the missing piece you are asking about. The signature of `zip` is `(...iterables) => zipped` where the signature of `apply(zip)` is `(iterables) => zipped`, which is what you'll receive in a pipeline.

I agree that it can be confusing and that people don't think to reach for `apply` when working in a pipeline, but I also think that people find `zip(a, b)` to be more natural to write than `zip([a, b])`.

**Reply by conartist6** [reply]
Of course it's also perfectly fine to do what you did and just use
`source => zip(source, map(transformValue, source))`
But either way the code isn't really correct (or it isn't **necessarily** correct) without that leading `toArray`!

**Reply by hillin** [reply]
Indeed, my `generate()` function creates an array so reusing it was not a matter. `toArray()` should be the correct thing to do. `apply(zip)` looks inspiring!
